### PR TITLE
Left ILU support

### PR DESCRIPTION
**Pull request by michel2323**
* I have to use `transpose!` for the factors for the ILU, as there's no constructor for `adjoint` or lazy transpose in `IncompleteLU.jl`.

### CODE REVIEW

File: src/EnzymeRules/enzymerules.jl [hunk 1]
```diff
@@ -155,7 +155,16 @@ for AMT in (:Matrix, :SparseMatrixCSC)
...
                 end
-                _b.dval .= Krylov.$solver(transpose(A), bx[]; M=M, N=N, verbose=verbose, options...)[1]
+                N = M
+                M = N
+                if isa(M, IncompleteLU.ILUFactorization)
+                    U = copy(M.U)
+                    L = copy(M.L)
+                    transpose!(U, M.L)
+                    transpose!(L, M.U)
+                    N = IncompleteLU.ILUFactorization(L, U)
```
Comment (amontoison) [inline src/EnzymeRules/enzymerules.jl:165]: You already permuted `M` and `N`. I think that it should be `M`.

File: src/EnzymeRules/enzymerules.jl [hunk 2]
```diff
@@ -155,7 +155,16 @@ for AMT in (:Matrix, :SparseMatrixCSC)
                 if verbose > 0
                     @info "($psolver, $pamt) reverse"
                 end
-                _b.dval .= Krylov.$solver(transpose(A), bx[]; M=M, N=N, verbose=verbose, options...)[1]
+                N = M
+                M = N
+                if isa(M, IncompleteLU.ILUFactorization)
```
Comment (amontoison) [inline src/EnzymeRules/enzymerules.jl]: You need to do the same thing if `N` is an `ILUFactorization` (right preconditioner).

File: src/EnzymeRules/enzymerules.jl (hunk 1 above)
Comment (michel2323) [reply src/EnzymeRules/enzymerules.jl:165]: Fixed

File: src/EnzymeRules/enzymerules.jl (hunk 2 above)
Comment (michel2323) [reply src/EnzymeRules/enzymerules.jl]: I'll do this in another PR.

**Pull request merged**